feat: refactor activities

djaiss opened the pull request:
This PR changes how activities are managed (using services) and displayed (now in Vue, making them inline).

Now that we are using services for that, it'll easier to add more functionalities to activities.

2 breaking changes (for now): 

- the API methods to manage activities are DISABLED and do not work anymore.
- it's not possible yet to update an existing activity

We'll do it in a future iteration so this PR doesn't get too big (it's already way too big).

lydgate commented:
Hi, is this still going to happen? I was running on this branch to fix some of the activities, just wondering if it will get merged, or if I'll need to do something to my DB to switch back to master.

djaiss commented:
> Hi, is this still going to happen? I was running on this branch to fix some of the activities, just wondering if it will get merged, or if I'll need to do something to my DB to switch back to master.

It will eventually get merged. As soon as I have time to finish it, but yes, it's coming.

djaiss commented:
Ready for review @asbiin

asbiin commented on CHANGELOG:
```diff
@@ -16,7 +17,7 @@ Fixes:
 * Fix contact missing create form
 * Fix money format for non two "2" minor unit currencies
-
+* Fix dashboard do not crash anymore when trying to get property contact of non-object
```
I've already rewrote this in 'Fix dashboard crash when reminder is empty' above

asbiin commented on app/Http/Controllers/Contacts/ActivitiesController.php:
```diff
@@ -23,12 +33,153 @@ public function __construct(ActivityStatisticService $service)
+
+        try {
+            app(DestroyActivity::class)->execute($data);
+        } catch (\Exception $e) {
```
You should use `ModelNotFoundException`

djaiss replied:
Indeed.

asbiin commented on app/Http/Controllers/Contacts/ActivitiesController.php:
```diff
@@ -23,12 +33,153 @@ public function __construct(ActivityStatisticService $service)
+        // also push the current contact
+        array_push($arrayParticipants, $contact->id);
+
+        $activity = app(AttachContactToActivity::class)->execute([
```
No try/catch here ?

djaiss replied:
Good point.

asbiin commented on app/Models/Account/Activity.php:
```diff
@@ -117,9 +125,9 @@ public function getSummary()
+    public function getDescriptionAttribute($value)
     {
-        return $this->description;
+        return $value;
```
This function is not needed anymore

djaiss replied:
You are right.

asbiin commented on app/Services/Account/Activity/Activity/AttachContactToActivity.php:
```diff
@@ -0,0 +1,66 @@
+        // reset current associations
+        $activity->contacts()->sync([]);
+
+        foreach ($data['contacts'] as &$contactId) {
```
Why using `&$contactId`?

asbiin commented on app/Services/Account/Activity/Activity/CreateActivity.php:
```diff
@@ -0,0 +1,81 @@
+            'activity_type_id' => 'nullable|integer',
+            'summary' => 'required|string:255',
+            'description' => 'nullable|string:400000000',
+            'date' => 'required|date_format:Y-m-d',
```
You may add `date` as a format
```suggestion
            'date' => 'required|date|date_format:Y-m-d',
```

djaiss replied:
Is it necessary though?

asbiin replied:
No I don't think so either

asbiin commented on app/Services/Account/Activity/Activity/CreateActivity.php:
```diff
@@ -0,0 +1,81 @@
+        // Log a journal entry
+        (new JournalEntry)->add($activity);
+
+        return Activity::find($activity->id);
```
Why not just return `$activity`? ( maybe after a refresh())

djaiss replied:
@asbiin Because `create` returns an object with only the fields that have been modified, whereas this line returns the object with all the attributes. I don't know what `refresh()` does.

asbiin replied:
From [Lararavel documentation](https://laravel.com/docs/5.8/eloquent#retrieving-models):
```
The refresh method will re-hydrate the existing model using fresh data from the database. In addition, all of its loaded relationships will be refreshed as well:
```
So it will do almost the same, but `$activity->refresh();` is more elegant, I think.

asbiin commented on app/Services/Account/Activity/Activity/UpdateActivity.php:
```diff
@@ -0,0 +1,53 @@
+            'activity_id' => 'required|integer|exists:activities,id',
+            'summary' => 'required|string:255',
+            'description' => 'nullable|string:400000000',
+            'date' => 'required|date_format:Y-m-d',
```
Add date format
```suggestion
            'date' => 'required|date|date_format:Y-m-d',
```

asbiin replied:
This is still todo right?

asbiin commented on resources/assets/js/components/people/activity/ActivityList.vue:
```diff
@@ -0,0 +1,196 @@
+          <a v-if="displayLogActivity == false" class="btn edit-information" @click="displayLogActivity = true">
+            {{ $t('people.activities_add_activity') }}
+          </a>
+          <a v-if="displayLogActivity" class="btn edit-information" @click="displayLogActivity = false">
```
```suggestion
          <a v-else class="btn edit-information" @click="displayLogActivity = false">
```

djaiss replied:
Good catch.

asbiin commented on resources/assets/js/components/people/activity/ActivityList.vue:
```diff
@@ -0,0 +1,196 @@
+        {{ activity.summary }}
+      </h2>
+
+      <div v-if="activity.description" class="pl2 pr2 pb3" v-html="activity.description">
```
`v-html` can lead to XSS vulnerability, are you sure you want to use it?

djaiss replied:
@asbiin description can contain markdown which needs to be rendered. There is no other way to display Markdown then.... OR I have to parse markdown in Vue and not in the backend. What do you think?

asbiin replied:
Oh well, is description is parsed on the backend, it should be OK.
Anyways, I'm not sure, but we could fix this (it's used in other part of the application) later.

asbiin commented on resources/assets/js/components/people/activity/ActivityList.vue:
```diff
@@ -0,0 +1,196 @@
+
+            <!-- PARTICIPANT LIST -->
+            <li v-if="activity.attendees.total > 1" class="di">
+              <ul :class="[ dirltr ? 'mr3 di list' : 'ml3 di list' ]">
```
```suggestion
              <ul class="di list" :class="[ dirltr ? 'mr3' : 'ml3' ]">
```

asbiin commented on resources/assets/js/components/people/activity/ActivityList.vue:
```diff
@@ -0,0 +1,196 @@
+        <div class="w-70" :class="[ dirltr ? 'fl' : 'fr' ]">
+          <ul class="list">
+            <!-- HAPPENED AT -->
+            <li :class="[ dirltr ? 'mr3 di' : 'ml3 di' ]">
```
```suggestion
            <li class="di" :class="[ dirltr ? 'mr3' : 'ml3' ]">
```

asbiin commented on resources/assets/js/components/people/activity/ActivityList.vue:
```diff
@@ -0,0 +1,196 @@
+
+            <!-- EMOTIONS LIST -->
+            <li v-if="activity.emotions.length != 0" class="di">
+              <ul :class="[ dirltr ? 'mr3 di list' : 'ml3 di list' ]">
```
```suggestion
              <ul class="di list" :class="[ dirltr ? 'mr3' : 'ml3' ]">
```

asbiin commented on resources/assets/js/components/people/activity/ActivityList.vue:
```diff
@@ -0,0 +1,196 @@
+            </li>
+
+            <!-- ACTIVITY TYPE -->
+            <li v-if="activity.activity_type" :class="[ dirltr ? 'mr3 di' : 'ml3 di' ]">
```
```suggestion
            <li v-if="activity.activity_type" class="di" :class="[ dirltr ? 'mr3' : 'ml3' ]">
```

asbiin commented on resources/assets/js/components/people/activity/ActivityList.vue:
```diff
@@ -0,0 +1,196 @@
+
+  methods: {
+    prepareComponent() {
+      this.dirltr = this.$root.htmldir == 'ltr';
```
This may be replaced with 
```
  computed: {
    dirltr() {
      return this.$root.htmldir == 'ltr';
    }
  },
```

asbiin commented on resources/assets/js/components/people/activity/ActivityTypeList.vue:
```diff
@@ -0,0 +1,47 @@
+
+  methods: {
+    prepareComponent(hash) {
+      this.dirltr = this.$root.htmldir == 'ltr';
```
This may be replaced with 
```
  computed: {
    dirltr() {
      return this.$root.htmldir == 'ltr';
    }
  },
```

asbiin commented on resources/assets/js/components/people/activity/ActivityTypeList.vue:
```diff
@@ -0,0 +1,47 @@
+  },
+
+  methods: {
+    prepareComponent(hash) {
```
`hash` is not used.

asbiin commented on resources/assets/js/components/people/activity/CreateActivity.vue:
```diff
@@ -0,0 +1,247 @@
+                <form-date
+                  v-model="newActivity.happened_at"
+                  :default-date="todayDate"
+                  :locale="'en'"
```
You missed the real locale.

asbiin commented on resources/assets/js/components/people/activity/CreateActivity.vue:
```diff
@@ -0,0 +1,247 @@
+        <div v-show="!displayDescription || !displayEmotions || !displayCategory" class="bb b--gray-monica pv3 mb3">
+          <ul class="list">
+            <li v-show="!displayDescription" class="di pointer mr3">
+              <a @click="displayDescription = true">Add more details</a>
```
i18n?

asbiin replied:
This is still todo

asbiin commented on resources/assets/js/components/people/activity/ActivityList.vue:
```diff
@@ -0,0 +1,196 @@
+      <h3 class="mb2">
+        🍿 {{ $t('people.activity_title') }}
+
+        <span class="fr relative" style="top: -7px;">
```
It should be better to add a claa (scoped) in the vue file.

djaiss commented:
@asbiin I just updated this, merged master to it and fixed the new bugs. I think it’s ready. The only things are 2 breaking changes (for now):

- the API methods to manage activities are DISABLED and do not work anymore.
- it's not possible yet to update an existing activity

I'll write the API methods once this PR is merged, as well as updating activities.

asbiin commented on app/Http/Controllers/Contacts/ActivitiesController.php:
```diff
@@ -23,12 +34,157 @@ public function __construct(ActivityStatisticService $service)
+        ]);
+
+        $arrayParticipants = [];
+        foreach ($request->get('participants') as $participant) {
```
use
```suggestion
        foreach ($request->input('participants') as $participant) {
```
